source/s3_compute_platform_info.c

```diff
@@ -0,0 +1,591 @@
+
+    s_add_platform_info_to_table(loader, &s_c5n_18xlarge_platform_info);
+    s_add_platform_info_to_table(loader, &s_c5n_9xlarge_platform_info);
+    s_add_platform_info_to_table(loader, &s_c5n_metal_platform_info);
+    s_add_platform_info_to_table(loader, &s_p4d_platform_info);
+    s_add_platform_info_to_table(loader, &s_p4de_platform_info);
+    s_add_platform_info_to_table(loader, &s_p5_platform_info);
+    s_add_platform_info_to_table(loader, &s_trn1_n_platform_info);
+    s_add_platform_info_to_table(loader, &s_trn1_platform_info);
+
+    return loader;
+}
+
+struct aws_s3_compute_platform_info_loader *aws_s3_compute_platform_info_loader_acquire(
+    struct aws_s3_compute_platform_info_loader *loader) {
+    aws_ref_count_acquire(&loader->ref_count);
+    return loader;
+}
+
+struct aws_s3_compute_platform_info_loader *aws_s3_compute_platform_info_loader_release(
+    struct aws_s3_compute_platform_info_loader *loader) {
+    if (loader) {
+        aws_ref_count_release(&loader->ref_count);
+    }
+    return NULL;
+}
+
+struct imds_callback_info {
+    struct aws_allocator *allocator;
+    struct aws_string *instance_type;
+    struct aws_condition_variable c_var;
+    int error_code;
+    bool shutdown_completed;
+    struct aws_mutex mutex;
+};
+
+static void s_imds_client_shutdown_completed(void *user_data) {
+    struct imds_callback_info *info = user_data;
+    aws_mutex_lock(&info->mutex);
+    info->shutdown_completed = true;
+    aws_mutex_unlock(&info->mutex);
+    aws_condition_variable_notify_all(&info->c_var);
+}
+
+static bool s_client_shutdown_predicate(void *arg) {
+    struct imds_callback_info *info = arg;
+    return info->shutdown_completed;
+}
+
+static void s_imds_client_on_get_instance_info_callback(
+    const struct aws_imds_instance_info *instance_info,
+    int error_code,
+    void *user_data) {
+    struct imds_callback_info *info = user_data;
+
+    aws_mutex_lock(&info->mutex);
+    if (error_code) {
+        info->error_code = error_code;
+    } else {
+        info->instance_type = aws_string_new_from_cursor(info->allocator, &instance_info->instance_type);
+    }
+    aws_mutex_unlock(&info->mutex);
+    aws_condition_variable_notify_all(&info->c_var);
+}
+
+static bool s_completion_predicate(void *arg) {
+    struct imds_callback_info *info = arg;
+    return info->error_code != 0 || info->instance_type != NULL;
+}
+
+struct aws_byte_cursor aws_s3_get_ec2_instance_type(struct aws_s3_compute_platform_info_loader *loader) {
+    aws_mutex_lock(&loader->lock_data.lock);
+    struct aws_byte_cursor return_cur;
+    AWS_ZERO_STRUCT(return_cur);
+
+    if (loader->lock_data.detected_instance_type) {
+        AWS_LOGF_TRACE(
+            AWS_LS_S3_CLIENT,
+            "id=%p: Instance type has already been determined to be %s. Returning cached version.",
+            (void *)loader,
+            aws_string_bytes(loader->lock_data.detected_instance_type));
+        goto return_instance_and_unlock;
```

Comment:
debatable: should we remember if the IMDS lookup fails?

currently, if the lookup fails, then `detected_instance_type` will be NULL, and every time this function is called we'll redo it

I guess this doesn't super matter, except where I make another comment in this PR about maybe we should call this defensively anytime we're about to look in the table

Reply:
The next call to IMDS might succeed. I would say it's user error if they keep calling this function repeatedly, even though it's failing.